src/partition/repart.c

```diff
@@ -3969,59 +3970,87 @@ static int partition_populate_filesystem(Partition *p, const char *node, const S
+                return log_error_errno(dfd, "Failed to stat source '%s%s': %m", strempty(arg_root), path);
+
+        r = readdir_all(dfd, 0, &de);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read '%s%s' contents: %m", strempty(arg_root), path);
+
+        for (size_t i = 0; i < de->n_entries; i++) {
+                _cleanup_free_ char *d = NULL;
+                struct stat st;
+
+                if (fstatat(dfd, de->entries[i]->d_name, &st, AT_SYMLINK_NOFOLLOW) < 0)
+                        return log_error_errno(r, "Failed to stat '%s': %m", de->entries[i]->d_name);
+
+                if (set_contains(*denylist, &st))
+                        continue;
+
+                d = memdup(&st, sizeof(st));
+                if (!d)
+                        return log_oom();
+                if (set_ensure_put(denylist, &inode_hash_ops, d) < 0)
+                        return log_oom();
+
+                TAKE_PTR(d);
+        }
+
+        return 0;
+}
+
+static int make_copy_files_denylist(Context *context, Partition *p, Set **ret) {
         _cleanup_set_free_ Set *denylist = NULL;
         int r;
 
         assert(context);
         assert(ret);
 
-        LIST_FOREACH(partitions, p, context->partitions) {
-                const char *sources = gpt_partition_type_mountpoint_nulstr(p->type);
+        LIST_FOREACH(partitions, q, context->partitions) {
+                if (p == q)
+                        continue;
+
+                const char *sources = gpt_partition_type_mountpoint_nulstr(q->type);
                 if (!sources)
                         continue;
 
                 NULSTR_FOREACH(s, sources) {
-                        _cleanup_free_ char *d = NULL;
-                        struct stat st;
+                        /* Exclude the children of partition mount points so that the nested partition mount
+                         * point itself still ends up in the upper partition. */
 
-                        r = chase_symlinks_and_stat(s, arg_root, CHASE_PREFIX_ROOT, NULL, &st, NULL);
-                        if (r == -ENOENT)
-                                continue;
+                        r = exclude_children(s, &denylist);
                         if (r < 0)
-                                return log_error_errno(r, "Failed to stat source file '%s%s': %m",
-                                                       strempty(arg_root), s);
-
-                        if (set_contains(denylist, &st))
-                                continue;
-
-                        d = memdup(&st, sizeof(st));
-                        if (!d)
-                                return log_oom();
-                        if (set_ensure_put(&denylist, &inode_hash_ops, d) < 0)
-                                return log_oom();
-
-                        TAKE_PTR(d);
+                                return r;
                 }
         }
 
+        FOREACH_STRING(s, "proc", "sys", "dev", "run", "tmp", "var/tmp") {
```

Comment:
var/tmp but not var?

Reply:
var should probably be excluded explicitly using `ExcludeFiles=`. There's too much stuff that still puts stuff in there during installation time that it seems like excluding it by default would probably cause issues.

Reply:
Also, users might simply want to pack up images with state and we shouldn't make that impossible with repart